server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/CassandraMigrationRoutes.java

```diff
@@ -72,15 +72,13 @@ public void define(Service service) {
                 response.status(NO_CONTENT);
             } catch (NullPointerException | IllegalArgumentException e) {
                 LOGGER.info("Invalid request for version upgrade");
-                response.status(INVALID_VERSION);
+                throw halt(HttpStatus.BAD_REQUEST_400, "Invalid request for version upgrade " + e.getMessage());
```

Comment:
never send the exception's message to the client because it could be a security risk, instead you can log it for later debugging

Reply:
The client here is necessarily an admin, with a **master, kill the world JWT**. This makes your assumption a bit weaker, IMO.

I would value a more explicit API, and I believe such security issues are not applicable.

Remember that your implementation strategy is to proxy webadmin for normal users, thus Admin OpenPaaS modules can filter that informations out.

I believe your **user <-> server** framework could be upgraded to **inter service communication**, and the considerations are not the same.

Food for thoughts.